qhub/cost.py

```diff
@@ -0,0 +1,112 @@
+    """
+    try:
+        subprocess.check_output(["infracost", "configure", "get", "api_key"])
+        return True
+    except subprocess.CalledProcessError:
+        return False
+
+
+def _run_infracost(path):
+    """
+    Run infracost on the given path and return the JSON output
+    """
+    try:
+        process = subprocess.Popen(
+            ["infracost", "breakdown", "--path", path, "--format", "json"],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        stdout, stderr = process.communicate()
+        infracost_data = json.loads(
+            re.search("({.+})", stdout.decode("UTF-8"))
+            .group(0)
+            .replace("u'", '"')
+            .replace("'", '"')
+        )
+        return infracost_data
+    except subprocess.CalledProcessError:
+        return None
+
+
+def infracost_report(path):
+    """
+    Generate a report of the infracost cost of the given path
+    args:
+        path: path to the qhub stages directory
+    """
+    if not path:
+        path = os.path.join(os.getcwd(), "stages")
+
+    if _check_infracost() and _check_infracost_api_key():
+        if not os.path.exists(path):
+            print("Deployment is not available")
+        else:
+            data = _run_infracost(path)
+
+            cost_table = Table(title="Cost Breakdown")
+            cost_table.add_column("Name", justify="right", style="cyan", no_wrap=True)
+            cost_table.add_column("Cost", justify="right", style="cyan", no_wrap=True)
+
+            cost_table.add_row("Total Monthly Cost", data["totalMonthlyCost"])
+            cost_table.add_row("Total Hourly Cost", data["totalHourlyCost"])
+
+            resource_table = Table(title="Resource Breakdown")
+            resource_table.add_column(
+                "Name", justify="right", style="cyan", no_wrap=True
+            )
+            resource_table.add_column(
+                "Number", justify="right", style="cyan", no_wrap=True
+            )
+
+            resource_table.add_row(
+                "Total Detected Costs", str(data["summary"]["totalDetectedResources"])
+            )
+            resource_table.add_row(
+                "Total Supported Resources",
+                str(data["summary"]["totalSupportedResources"]),
+            )
+            resource_table.add_row(
+                "Total Un-Supported Resources",
+                str(data["summary"]["totalUnsupportedResources"]),
+            )
+            resource_table.add_row(
+                "Total Non-Priced Resources",
+                str(data["summary"]["totalNoPriceResources"]),
+            )
+            resource_table.add_row(
+                "Total Usage-Priced Resources",
+                str(data["summary"]["totalUsageBasedResources"]),
+            )
+
+            console = Console()
+            console.print(cost_table)
+            console.print(resource_table)
+            print(f"Access the dashboard here: {data['shareUrl']}")
```

Comment:
@HarshCasper when I trying this out myself, I kept running into this issue:

```
Traceback (most recent call last):
  File "/Users/eskild/opt/miniconda3/envs/qhub-dev/bin/qhub", line 33, in <module>
    sys.exit(load_entry_point('qhub', 'console_scripts', 'qhub')())
  File "/Users/eskild/Documents/Quansight/qhub/qhub/__main__.py", line 7, in main
    cli(sys.argv[1:])
  File "/Users/eskild/Documents/Quansight/qhub/qhub/cli/__init__.py", line 55, in cli
    args.func(args)
  File "/Users/eskild/Documents/Quansight/qhub/qhub/cli/cost.py", line 15, in handle_cost_report
    infracost_report(args.path)
  File "/Users/eskild/Documents/Quansight/qhub/qhub/cost.py", line 110, in infracost_report
    print(f"Access the dashboard here: {data['shareUrl']}")
KeyError: 'shareUrl'
```

Would you mind having a look?

Reply:
FWIW, I was trying this locally, trying to get an estimate for `~/Documents/Quansight/qhub-deployment-quansight-beta`

Reply:
Hi @iameskild 👋 

I tried running it and check what the error was. Fortunately it worked for me:

```sh
       Cost Breakdown        
┏━━━━━━━━━━━━━━━━━━━━┳━━━━━━┓
┃               Name ┃ Cost ┃
┡━━━━━━━━━━━━━━━━━━━━╇━━━━━━┩
│ Total Monthly Cost │   73 │
│  Total Hourly Cost │  0.1 │
└────────────────────┴──────┘
           Resource Breakdown            
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓
┃                         Name ┃ Number ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩
│         Total Detected Costs │     11 │
│    Total Supported Resources │      3 │
│ Total Un-Supported Resources │      1 │
│   Total Non-Priced Resources │      7 │
│ Total Usage-Priced Resources │      3 │
└──────────────────────────────┴────────┘
Access the dashboard here: https://dashboard.infracost.io/share/cuy3fqeeb3kqvcufrme7c9hlekcnuxa7
```

Reply:
Can you please run these two commands and see what the output is:

```sh
infracost --version
infracost configure get api_key
```

Reply:
Hi @iameskild — I took a look at it and found the root cause. The JSON payload that we received did not have dashboards enabled in it, hence QHub errored out. 

To enable this, we have to enable the dashboards via:

```sh
infracost configure set enable_dashboard true
```

Can you run this and try it out once? Since we are giving the users control over their Infracost installation, I suggest we add it out in the docs to let the users know to enable dashboards. Or we can add it as a flag that the user needs to add in case they wish to use it. What do you think? cc: @viniciusdc

Reply:
Thanks for the update @HarshCasper! I tried it with `enable_dashboard true` and it worked. The dashboard link is very neat!

I think you bring up a good point about how much of the `infracost` setup process we should leave to the user. I'm inclined to have this tool handle the setup for the user. Handling the install with a `qhub cost-estimate` subcommand might be a too much to add right now but adding the flag to enable dashboards sounds like a reasonable addition.

And like you said, either way, we should document this :)

Reply:
Added it 👍

Reply:
I agree with @iameskild, we should let the user do the config by themselves which restricts the maintenance of this in the future.